### PR TITLE
Players

### PR DESCRIPTION
**Pull request by Middlebrooks314**
Gives the human players the ability to choose a mark and use it in the game. 

https://trello.com/c/Qib1RlEn

### CODE REVIEW

File: spec/game_spec.rb [hunk 1]
```diff
@@ -1,260 +0,0 @@
-require 'spec_helper'
```
Comment (cmvandrevala) [inline spec/game_spec.rb:1]: Why did you delete all of the game specs? Don't you test `play` and a few other functions in `Game`?

File: lib/player_creator.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+require 'pry'
```
Comment (cmvandrevala) [inline lib/player_creator.rb]: Remove `require 'pry'` as it is only used for debugging

File: lib/player_creator.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+class PlayerCreator
```
Comment (cmvandrevala) [inline lib/player_creator.rb]: So, right now you pass an instance of `PlayerCreator` into your `Game` class. IMO, this is a bit backwards. I might have `PlayerCreator` return a new instance of `Game` rather than the other way around. This code is probably fine for this particular story, but you might run into architectural issues when you implement computer players

Comment (cmvandrevala) [reply lib/player_creator.rb]: No need to act on this comment if you don't agree with me, BTW. Just want to push back on this design decision.

File: lib/human_player.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    @mark = mark
+  end
+
+  def move
+    @display.make_move_message(@mark)
+    index = @display.retrieve_user_input.to_i - 1
+    @board.valid_move?(index) ? index : move
+  end
+
+  def create_mark(other_mark = nil)
```
Comment (cmvandrevala) [inline lib/human_player.rb]: This method should probably be located in `PlayerCreator`. Like, determine which mark a player needs and then initialize a new player with that mark. Otherwise, how will you specify a mark for a computer player?

File: lib/human_player.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    @display.setup_mark_message
+    potential_mark = @display.retrieve_user_input
+    mark = check_game_mark(potential_mark, other_mark)
+    @mark = mark
+  end
+
+  def check_game_mark(potential_mark, other_mark)
```
Comment (cmvandrevala) [inline lib/human_player.rb]: Can this be a private method?

File: lib/game.rb [hunk 6]
```diff
@@ -1,56 +1,59 @@
...
+
+  private
+
+  def switch_players
+    if !rules.game_over?(board)
+      @current_player == @player_one ? @current_player = @player_two : @current_player = @player_one
+    end
   end
 
   def winner
```
Comment (cmvandrevala) [inline lib/game.rb:46]: Ruby prefers a slightly different convention:

```
def winner
  @current_player if rules.winning_combination?(board)
end
```

File: lib/game.rb [hunk 7]
```diff
@@ -1,56 +1,59 @@
...
-      @current_player == 'X' ? @current_player = 'O' : @current_player = 'X'
-    end
+  def player_setup
+    @player_creator.create_players(@board, @display)
+    @player_one = @player_creator.player_one
+    @player_two = @player_creator.player_two
   end
 
   def play
+    @current_player = @player_one
```
Comment (cmvandrevala) [inline lib/game.rb]: Why do you need this line? Isn't `@current_player` specified in the initializer?

File: lib/game.rb [hunk 8]
```diff
@@ -1,56 +1,59 @@
 require_relative 'board'
-require_relative 'console'
 require_relative 'rules'
+require_relative 'human_player'
+require_relative 'dumb_player'
```
Comment (cmvandrevala) [inline lib/game.rb]: The `dumb_player` was just an example that we used during our pairing session to prove that the polymorphism works. You can probably delete that class

File: lib/dumb_player.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+class DumbPlayer
```
Comment (cmvandrevala) [inline lib/dumb_player.rb]: You can probably delete this now

**Review by cmvandrevala — CHANGES_REQUESTED** [review]

File: lib/display.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
+class Display
+
```
Comment (aaizuss) [inline lib/display.rb]: Can you delete this line? And I'm glad you put these in a class!

File: lib/game.rb [hunk 11]
```diff
@@ -1,56 +1,59 @@
...
 
-  def initialize(board, console, rules)
+  def initialize(board, rules, display, player_creator)
     @board = board
-    @console = console
-    @current_player = 'X'
     @rules = rules
+    @display = display
+    @player_creator = player_creator
+    @current_player = @player_one
```
Comment (aaizuss) [inline lib/game.rb:13]: It's strange that `@player_one` is used here, but it isn't actually set until the `player_setup` method gets called in `start`

File: lib/game.rb (hunk 7 above)
Comment (Middlebrooks314) [reply lib/game.rb]: When the game is initialized the one in the initialize method is = nil. Because Ruby passes by value and not by reference, then it needs to be set again. I have deleted it from the initialize method now. Any reason this isnt a good idea?

File: lib/game.rb (hunk 11 above)
Comment (Middlebrooks314) [reply lib/game.rb]: Cyrus mentioned this as well. I removed it from the initialize method and kept it in the play method because you're right, it initializes as nil.

File: spec/game_spec.rb (hunk 1 above)
Comment (Middlebrooks314) [reply spec/game_spec.rb:1]: I was confused, I thought all of the methods were being tested with each of the unit tests in their respective class specs. I talked to Josh about integration tests today and I reimplemented them.

File: bin/run.rb [hunk 12]
```diff
@@ -1,9 +1,9 @@
-require_relative '../lib/game.rb'
-require_relative '../lib/console.rb'
+require 'require_all'
+require_all 'lib'
```
Comment (cmvandrevala) [inline bin/run.rb:2]: I had no idea that this gem existed. This is super cool

File: lib/board.rb [hunk 13]
```diff
@@ -1,5 +1,5 @@
 class Board
-  attr_accessor :squares
+  attr_reader :squares
```
Comment (cmvandrevala) [inline lib/board.rb:2]: Nice

**Review by cmvandrevala — APPROVED** [review]

File: lib/game.rb [hunk 14]
```diff
@@ -1,56 +1,49 @@
...
-    board.valid_move?(index) ? @board.mark(index, @current_player) : retrieve_user_move
-  end
-
-  def switch_players
-    if !rules.game_over?(board)
-      @current_player == 'X' ? @current_player = 'O' : @current_player = 'X'
-    end
-  end
-
   def play
```
Comment (aaizuss) [inline lib/game.rb:28]: this reads really nicely! I'd move the method higher up (above `turn`) since it uses `turn`. It's nice to have the higher level methods towards the top of the file, and keep the helper methods below them.

File: lib/game_factory.rb [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+      potential_mark = @display.retrieve_user_input
+    end
+    potential_mark
+  end
+
+  def valid_game_piece?(potential_mark, other_mark)
+    if potential_mark.length > 1 || potential_mark == ' ' || potential_mark == '' || potential_mark == other_mark
+      false
+    else
+      true
```
Comment (aaizuss) [inline lib/game_factory.rb:44]: First, you have `potential_mark == ''` twice.
You can save a few lines by writing the above as
```ruby
def valid_game_piece?(potential_mark, other_mark)
  true unless potential_mark.length > 1 || potential_mark == ' ' || potential_mark == other_mark
end
```

But here's a tip for functions that check a condition and then return `true` or `false` - you don't have to explicitly return `true`/`false`

I usually try to avoid explicitly returning true/false when the answer is completely based on a condition, but that's a little harder to do here since your conditions are written for the `false` case. 

Given the boolean expressions that you've written, you can have a one line function if you change your function name to be `invalid` game piece, and if you want to use this, you'll need to change your `check_game_mark` method.
```ruby
def invalid_game_piece?(potential_mark, other_mark)
  potential_mark.length > 1 || potential_mark == ' ' || potential_mark == other_mark
end
```

Or, if you want this function to continue to be called `valid` (and not change how your `check_game_mark` method works), we need to think of what conditions are necessary for the piece to be **valid** (since currently you've written the conditions for the invalid case)

I don't remember if we said that game marks could be more than one character, but assuming they need to be one character:
```ruby
def valid_game_piece?(potential_mark, other_mark)
  potential_mark.length == 1 && potential_mark != ' '  && potential_mark != other_mark
end
```
Now that I've written this out, I realize that `potential_mark.length == 1 && potential_mark != ' '` are actually redundant - if we restrict the length to 1, then the mark `''` won't be valid, and we could remove that condition, leaving us with 

```ruby
def valid_game_piece?(potential_mark, other_mark)
  potential_mark.length == 1  && potential_mark != other_mark
end
```

😄

**Review by aaizuss — APPROVED** [review]
I hope you're proud of how your code looks! I'm leaving some comments/suggestions, but they are not merge-blocking. Feel free to implement my suggestions in your next PR if you want

File: lib/rules.rb [hunk 16]
```diff
@@ -14,19 +17,19 @@ class Rules
...
+      @board.squares[x] == @board.squares[y] &&
+      @board.squares[x] == @board.squares[z]
     end
   end
 
-  def check_for_draw?(board)
-    winning_combination?(board) == false && board.full? == true
+  def check_for_draw?
+    winning_combination? == false && @board.full? == true
   end
```
Comment (aaizuss) [inline lib/rules.rb:34]: Since this method just returns a boolean, you could name it `draw?` (I think `check` is redundant - this is subjective though. It's not a bad name!)

More importantly, you don't need to explicitly compare your booleans to `true`/`false` - that *is* redundant. The body of this method can be
`!winning_combination? && @board.full?`

File: spec/display_spec.rb [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+      board = Board.new
+      display = Display.new(console, board)
+      display.welcome
+
+      expect(console.printed_string).to eq('Welcome to Tic Tac Toe')
+    end
+  end
+
+  describe 'make_move_message' do
+    it 'asks the current player for their move' do
```
Comment (aaizuss) [inline spec/display_spec.rb:18]: nitpicky, but I'd say that it asks the **given player** for their move

File: spec/display_spec.rb [hunk 18]
```diff
@@ -0,0 +1,96 @@
+require 'require_all'
+require_all 'lib'
+require_relative './fake_console.rb'
+
+describe Display do
```
Comment (aaizuss) [inline spec/display_spec.rb:5]: You might want to look into using `let` or a `subject` so you don't have to initialize `Display` the same way for every test. Here's a good explanation https://stackoverflow.com/questions/38437162/whats-the-difference-between-rspecs-subject-and-let-when-should-they-be-used

File: lib/rules.rb [hunk 19]
```diff
@@ -14,19 +17,19 @@ class Rules
     [2, 4, 6]
 ]
 
-  def game_over?(board)
-    winning_combination?(board) || check_for_draw?(board)
+  def game_over?
```
Comment (aaizuss) [inline lib/rules.rb:20]: Why did you decide to remove the `board` argument? (I see that you moved it to an instance variable, but why?)

**Review by aaizuss — COMMENTED** [review]
More feedback - again, don't worry about making these changes now - if you want to take the suggestions, you can make them after merging

File: lib/game.rb (hunk 14 above)
Comment (Middlebrooks314) [reply lib/game.rb:28]: Good to know! In next PR switch_players is private and retrieve_user_mark doesnt exist anymore, but still good to know.

File: lib/game_factory.rb (hunk 15 above)
Comment (Middlebrooks314) [reply lib/game_factory.rb:44]: This was a very good code walkthrough!

File: lib/rules.rb (hunk 19 above)
Comment (Middlebrooks314) [reply lib/rules.rb:20]: I felt like my code was littered with passing board arguments. After you asked the question I was wondering if I unnecessarily exposed the board? 🤔 💭
